kernel/kernel-api/src/main/java/io/delta/kernel/Transaction.java

```
@@ -0,0 +1,207 @@
+     * column name to partition value (as {@link Literal}). If the table is an un-partitioned table,
+     * then map should be empty.
+     *
+     * @param tableClient      {@link TableClient} instance to use.
+     * @param transactionState The transaction state
+     * @param dataIter         Iterator of logical data to transform to physical data. All the data
+     *                         in this iterator should belong to one physical partition.
+     * @param partitionValues  The partition values for the data. If the table is un-partitioned,
+     *                         the map should be empty
+     * @return Iterator of physical data to write to the data files.
+     */
+    static CloseableIterator<FilteredColumnarBatch> transformLogicalData(
+            TableClient tableClient,
+            Row transactionState,
+            CloseableIterator<FilteredColumnarBatch> dataIter,
+            Map<String, Literal> partitionValues) {
+        validatePartitionValues(
+                getPartitionColumnsList(transactionState), partitionValues);
+        // TODO: add support for:
+        // - enforcing the constraints
+        // - generating the default value columns
+        // - generating the generated columns
+
+        // Remove the partition columns from the data as they are already part of file metadata
+        // and are not needed in the data files. TODO: once we start supporting uniform complaint
+        //  tables, we may conditionally skip this step.
+
+        // TODO: set the correct schema once we start supporting writing into column mapping table.
+        return dataIter.map(
+                filteredColumnarBatch -> {
+                    ColumnarBatch data = filteredColumnarBatch.getData();
+                    for (String partitionColName : partitionValues.keySet()) {
+                        int partitionColIndex = data.getSchema().indexOf(partitionColName);
```

any checks / validation we can do here? any chance that the index is -1? can we throw a better error than index out of bounds?